types/react-redux/index.d.ts

```diff
@@ -80,17 +80,17 @@ export interface Connect {
 
     <TStateProps = {}, no_dispatch = {}, TOwnProps = {}, State = {}>(
         mapStateToProps: MapStateToPropsParam<TStateProps, TOwnProps, State>
-    ): InferableComponentEnhancerWithProps<TStateProps & DispatchProp<any>, TOwnProps>;
+    ): InferableComponentEnhancerWithProps<TStateProps & DispatchProp<any> & TOwnProps, TOwnProps>;
```

Comment:
So `InferableComponentEnhancerWithProps` is receiving `TOwnProps` twice now. I think it would be better to fix this within that interface? You mention that type inference is failing, but trying this out locally and it seems that inference is working properly. It looks like `InferableComponentEnhancerWithProps` is doing the wrong thing with `TOwnProps`.

Perhaps changing that interface to:
```typescript
export interface InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> {
    <P extends TInjectedProps>(
        component: Component<P & TNeedsProps> // this is the line that changed
    ): ComponentClass<Omit<P, keyof TInjectedProps> & TNeedsProps> & {WrappedComponent: Component<P>}
}
```

Reply:
Yup, you are right. Good call. Looking into `mergeProps` a bit, and we cannot assume that `OwnProps` will always be there when `mergeProps` is supplied.

Reply:
This is breaking our implementation.
We can no longer decorate component whose props don't extends TOwnProps.

It is a problem, because own props can be required by mapStateToProps and / or mapDispatchToProps and not by the decorated component.

It is leading to leaking state consumption logic into a component that shouldn't be aware of such implication.
I think we might have a contravariance conflict: it should be possible to decorate a component that ignores some of the props it receives, whether they are injected by mapState, mapDisptach or merge.